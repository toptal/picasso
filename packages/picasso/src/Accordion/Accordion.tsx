import React, {
  ReactNode,
  ChangeEvent,
  HTMLAttributes,
  ReactElement,
  forwardRef,
  useState,
} from 'react'
import cx from 'classnames'
import MUIAccordion from '@material-ui/core/Accordion'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { StandardProps, TransitionProps } from '@toptal/picasso-shared'

import { ArrowDownMinor16 } from '../Icon'
import AccordionSummary from '../AccordionSummary'
import AccordionDetails from '../AccordionDetails'
import styles from './styles'
import ButtonAction from '../ButtonAction'

export type Borders = 'all' | 'middle' | 'none'

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoAccordion',
})

const EmptyAccordionSummary = ({
  'data-testid': dataTestId,
}: {
  'data-testid'?: string
}) => <div data-testid={dataTestId} />

export interface Props
  extends StandardProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'onChange'> {
  /** Always visible part of accordion */
  children?: ReactNode
  /** Collapsible content of `Accordion` */
  content: ReactNode
  /** Define accordion content state, whether it should be collapsed or expanded */
  expanded?: boolean
  /** Define accordion initial content state */
  defaultExpanded?: boolean
  /** Whether the Accordion is disabled */
  disabled?: boolean
  /** Customize icon indicating expanded status */
  expandIcon?: ReactElement
  /** Defines where the horizontal borders show */
  borders?: Borders
  /** Callback invoked when `Accordion` item is toggled */
  onChange?: (event: ChangeEvent<{}>, expanded: boolean) => void
  /** Animation lifecycle callbacks. Backed by [react-transition-group/Transition](https://reactcommunity.org/react-transition-group/transition#Transition-props) */
  transitionProps?: TransitionProps
  testIds?: {
    emptyAccordionSummary?: string
    accordionSummary?: string
  }
}

const decorateWithExpandIconClasses = (
  expandIcon: ReactElement,
  classes: string
) =>
  React.cloneElement(expandIcon, {
    className: cx(expandIcon.props.className, classes),
  })

/* eslint-disable complexity */
export const Accordion = forwardRef<HTMLElement, Props>(function Accordion(
  props,
  ref
) {
  const {
    children,
    content,
    expanded,
    defaultExpanded,
    expandIcon,
    borders,
    disabled,
    className,
    style,
    onChange,
    testIds,
    transitionProps,
    ...rest
  } = props

  const classes = useStyles(props)
  const borderClasses: { [key in Borders]: string } = {
    all: classes.bordersAll,
    middle: classes.bordersMiddle,
    none: classes.bordersNone,
  }
  const [summaryExpanded, setSummaryExpanded] = useState(defaultExpanded)
  const [prevExpanded, setPrevExpanded] = useState(defaultExpanded)

  // getDerivedStateFromProps implementation to allow expanded to be controlled
  if (expanded !== undefined && expanded !== prevExpanded) {
    setSummaryExpanded(expanded)
    setPrevExpanded(expanded)
  }

  const handleSummaryClick = () => {
    setSummaryExpanded(!summaryExpanded)
  }

  const expandIconClass = cx(classes.expandIcon, {
    [classes.expandIconExpanded]: summaryExpanded,
  })

  const appliedBorders = children || expanded ? (borders as Borders) : 'none'

  return (
    <MUIAccordion
      {...rest}
      ref={ref}
      classes={{
        root: cx(classes.root, borderClasses[appliedBorders]),
      }}
      className={className}
      style={style}
      elevation={0}
      expanded={summaryExpanded}
      disabled={disabled}
      onChange={onChange}
      TransitionProps={transitionProps}
    >
      {children ? (
        <AccordionSummary
          classes={{
            root: classes.summary,
            content: classes.content,
          }}
          expandIcon={null}
          onClick={handleSummaryClick}
          data-testid={testIds?.accordionSummary}
        >
          {children}
          {expandIcon ? (
            decorateWithExpandIconClasses(expandIcon, expandIconClass)
          ) : (
            <div className={classes.expandIconAlignTop}>
              <ButtonAction
                icon={<ArrowDownMinor16 className={expandIconClass} />}
              />
            </div>
          )}
        </AccordionSummary>
      ) : (
        <EmptyAccordionSummary data-testid={testIds?.emptyAccordionSummary} />
      )}
      <AccordionDetails
        classes={{
          root: classes.details,
        }}
      >
        {content}
      </AccordionDetails>
    </MUIAccordion>
  )
})

Accordion.defaultProps = {
  borders: 'all',
  defaultExpanded: false,
  disabled: false,
  expanded: undefined,
  onChange: () => {},
}

Accordion.displayName = 'Accordion'

export default Accordion
