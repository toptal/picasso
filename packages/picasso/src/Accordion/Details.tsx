import React, { ReactNode } from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { StandardProps } from '@toptal/picasso-shared'

import styles from './styles'

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoAccordion',
})

export interface DetailsProps extends StandardProps {
  children: ReactNode
}

const Details = (props: DetailsProps) => {
  const { children, className, ...rest } = props
  const classes = useStyles(props)

  return (
    <div {...rest} className={cx(className, classes.detailsWrapper)}>
      {children}
    </div>
  )
}

export default Details
