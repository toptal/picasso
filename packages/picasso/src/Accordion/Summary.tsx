import React, { ReactNode } from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { StandardProps } from '@toptal/picasso-shared'

import styles from './styles'

export type Borders = 'all' | 'middle' | 'none'

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoAccordion',
})

export interface SummaryProps extends StandardProps {
  children: ReactNode
}

const Summary = (props: SummaryProps) => {
  const { children, className, ...rest } = props
  const classes = useStyles(props)

  return (
    <div {...rest} className={cx(className, classes.summaryWrapper)}>
      {children}
    </div>
  )
}

export default Summary
