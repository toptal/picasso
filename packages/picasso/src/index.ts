export { BaseProps } from '@toptal/picasso-shared'
export type { ColorType, SpacingType, SizeType } from '@toptal/picasso-shared'
export { AccordionCompound as Accordion } from './AccordionCompound'
export type { AccordionProps } from './Accordion'
export { default as AccountSelect } from './AccountSelect'
export type { AccountSelectProps } from './AccountSelect'
export { default as Amount } from './Amount'
export type { AmountProps } from './Amount'
export { default as Autocomplete } from './Autocomplete'
export type {
  AutocompleteProps,
  Item as AutocompleteItem,
} from './Autocomplete'
export { AvatarCompound as Avatar } from './AvatarCompound'
export type { AvatarProps } from './Avatar'
export { ButtonCompound as Button } from './ButtonCompound'
export type { ButtonProps, VariantType as ButtonVariantType } from './Button'
export type { ButtonCheckboxProps } from './ButtonCheckbox'
export type { ButtonRadioProps } from './ButtonRadio'
export { CheckboxCompound as Checkbox } from './CheckboxCompound'
export type { CheckboxProps } from './Checkbox'
export type { CheckboxGroupProps } from './CheckboxGroup'
export { default as Container } from './Container'
export type { ContainerProps } from './Container'
export { DropdownCompound as Dropdown } from './DropdownCompound'
export type { DropdownProps } from './Dropdown'
export { default as DropdownArrow } from './DropdownArrow'
export type { DropdownArrowProps } from './DropdownArrow'
export { default as EnvironmentBanner } from './EnvironmentBanner'
export type { EnvironmentBannerProps } from './EnvironmentBanner'
export { default as FileInput } from './FileInput'
export type { FileInputProps } from './FileInput'
export { default as FileList } from './FileList'
export type { FileListProps } from './FileList'
export { FormCompound as Form } from './FormCompound'
export type { FormProps } from './Form'
export { default as FormHint } from './FormHint'
export type { FormHintProps } from './FormHint'
export { default as FormError } from './FormError'
export type { FormErrorProps } from './FormError'
export { default as FormLabel } from './FormLabel'
export type { FormLabelProps, RequiredDecoration } from './FormLabel'
export { GridCompound as Grid } from './GridCompound'
export type { GridProps, GridSize } from './Grid'
export { HelpboxCompound as Helpbox } from './HelpboxCompound'
export type { HelpboxProps } from './Helpbox'
export { default as Image } from './Image'
export type { ImageProps } from './Image'
export { default as Indicator } from './Indicator'
export type { IndicatorProps } from './Indicator'
export { default as Input } from './Input'
export type { InputProps } from './Input'
export { default as InputAdornment } from './InputAdornment'
export type { InputAdornmentProps } from './InputAdornment'
export { TagCompound as Tag } from './TagCompound'
export type { TagProps } from './Tag'
export type { TagRectangularProps } from './TagRectangular'
export { default as Link } from './Link'
export type { LinkProps } from './Link'
export { default as Loader } from './Loader'
export type { LoaderProps } from './Loader'
export { default as Logo } from './Logo'
export type { LogoProps } from './Logo'
export { MenuCompound as Menu } from './MenuCompound'
export type { MenuProps } from './Menu'
export { default as MenuItem } from './MenuItem'
export type { MenuItemProps } from './MenuItem'
export { ModalCompound as Modal } from './ModalCompound'
export type { ModalProps } from './Modal'
export { default as MonthSelect } from './MonthSelect'
export type { MonthSelectProps } from './MonthSelect'
export { NotificationCompound as Notification } from './NotificationCompound'
export type { NotificationProps } from './Notification'
export { default as NotificationActions } from './NotificationActions'
export type { NotificationActionsProps } from './NotificationActions'
export { default as NumberInput } from './NumberInput'
export type { NumberInputProps } from './NumberInput'
export { PageCompound as Page } from './PageCompound'
export type { PageProps } from './Page'
export { default as Pagination } from './Pagination'
export type { PaginationProps } from './Pagination'
export { default as Paper } from './Paper'
export type { PaperProps } from './Paper'
export { default as PromptModal } from './PromptModal'
export type { PromptModalProps } from './PromptModal'
export { RadioCompound as Radio } from './RadioCompound'
export type { RadioProps } from './Radio'
export type { RadioGroupProps } from './RadioGroup'
export { default as RichTextEditor } from './RichTextEditor'
export type {
  RichTextEditorProps,
  RichTextEditorChangeHandler,
} from './RichTextEditor'
export { default as Select } from './Select'
export type { SelectProps } from './Select'
export type {
  Option as SelectOption,
  ValueType as SelectValueType,
} from './SelectBase'
export type { PageSidebarProps } from './PageSidebar'
export { default as Slider } from './Slider'
export type { SliderProps } from './Slider'
export { default as ShowMore } from './ShowMore'
export type { ShowMoreProps } from './ShowMore'
export { StepperCompound as Stepper } from './StepperCompound'
export type { StepperProps } from './Stepper'
export type { StepperVerticalProps } from './StepperVertical'
export { TableCompound as Table } from './TableCompound'
export type { TableProps } from './Table'
export { default as TableFooter } from './TableFooter'
export type { TableFooterProps } from './TableFooter'
export { TabsCompound as Tabs } from './TabsCompound'
export type { TabsProps } from './Tabs'
export { TagSelectorCompound as TagSelector } from './TagSelectorCompound'
export type { TagSelectorProps } from './TagSelector'
export { default as Tooltip } from './Tooltip'
export type { TooltipProps } from './Tooltip'
export { default as Typography } from './Typography'
export type { TypographyProps } from './Typography'
export { default as UserBadge } from './UserBadge'
export type { UserBadgeProps } from './UserBadge'
export { default as YearSelect } from './YearSelect'
export type { YearSelectProps } from './YearSelect'
export { default as ApplicationUpdateNotification } from './ApplicationUpdateNotification'
export type { ApplicationUpdateNotificationProps } from './ApplicationUpdateNotification'
export { AlertCompound as Alert } from './AlertCompound'
export type { AlertProps } from './Alert'
export type { AlertInlineProps } from './AlertInline'
export { default as Switch } from './Switch'
export type { SwitchProps } from './Switch'
export { default as Quote } from './Quote'
export type { QuoteProps } from './Quote'
export { default as Rating } from './Rating'
export type { RatingStarsProps, RatingThumbsProps } from './Rating'
export {
  default as TreeView,
  TreeViewContainer,
  useTreeView,
  TreeNodeAvatar,
} from './TreeView'
export type { TreeNodeInterface } from './TreeView'
export { default as StaticTreeView } from './StaticTreeView'
export { default as TimePicker } from './TimePicker'
export type { TimePickerProps } from './TimePicker'
export { default as SkeletonLoader } from './SkeletonLoader'
export { default as PageHead } from './PageHead'
export type { PageHeadProps } from './PageHead'
export { default as ProgressBar } from './ProgressBar'
export type { ProgressBarProps } from './ProgressBar'
export { default as List } from './List'
export type { ListProps } from './List'
export { default as Badge } from './Badge'
export type { BadgeProps } from './Badge'
export { BreadcrumbsCompound as Breadcrumbs } from './BreadcrumbsCompound'
export type { BreadcrumbsProps } from './Breadcrumbs'
export { default as Calendar } from './Calendar'
export type { DateOrDateRangeType } from './Calendar'
export {
  default as DatePicker,
  datePickerParseDateString,
  DEFAULT_DATE_PICKER_DISPLAY_DATE_FORMAT,
  DEFAULT_DATE_PICKER_EDIT_DATE_FORMAT,
} from './DatePicker'
export type {
  DatePickerValue,
  DatePickerInputCustomValueParser,
  DatePickerInputCustomValueParserParameters,
  DatePickerProps,
} from './DatePicker'
export { default as Drawer } from './Drawer'
export type { DrawerProps } from './Drawer'
export { default as Dropzone, ErrorCode as DropzoneErrorCode } from './Dropzone'
export type {
  DropzoneProps,
  FileUpload as DropzoneFileUpload,
  FileError as DropzoneFileError,
  FileRejection as DropzoneFileRejection,
  DropEvent as DropzoneDropEvent,
} from './Dropzone'
export { OverviewBlockCompound as OverviewBlock } from './OverviewBlockCompound'
export type { OverviewBlockProps } from './OverviewBlock'
export { default as TypographyOverflow } from './TypographyOverflow'
export type { TypographyOverflowProps } from './TypographyOverflow'
export { default as EmptyState } from './EmptyState'
export { default as Section } from './Section'
export type { SectionProps } from './Section'
export { NoteCompound as Note } from './NoteCompound'
export type { NoteProps } from './Note'
export { default as Timeline } from './Timeline'
export type { TimelineProps } from './Timeline'
export { default as RichText } from './RichText'
export type { RichTextProps, ASTType } from './RichText'
export { default as PasswordInput } from './PasswordInput'
export type { PasswordInputProps } from './PasswordInput'
export { default as FieldRequirements } from './FieldRequirements'
export type {
  FieldRequirementsProps,
  FieldRequirement,
} from './FieldRequirements'
export type { Status as OutlinedInputStatus } from './OutlinedInput'
export { default as AvatarUpload } from './AvatarUpload'
export type {
  AvatarUploadProps,
  FileUpload as AvatarUploadFileUpload,
  FileError as AvatarUploadFileError,
  FileRejection as AvatarUploadFileRejection,
  DropEvent as AvatarUploadDropEvent,
} from './AvatarUpload'

// hygen code generator inserts export statements above this comment.
export * from './Icon'
