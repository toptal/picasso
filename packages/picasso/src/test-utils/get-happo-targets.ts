type Result = Record<string, any>

/**
 * Generates Happo targets for responsive visual tests
 *
 * @param screenWidths Screen widths to generate targets for
 * @returns {Result} Object with Happo targets
 */
export const getHappoTargets = (screenWidths: number[]) =>
  screenWidths.reduce<Result>((acc: Result, width: number) => {
    const name = `chrome-desktop-width-${width}`

    acc[name] = {
      name,
      browser: 'chrome',
      viewport: `${width}x1024`,
    }

    return acc
  }, {})
