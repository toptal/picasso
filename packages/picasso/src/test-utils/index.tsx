import type { ReactElement } from 'react'
import React from 'react'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'

import TestingPicasso from './TestingPicasso'

export type PicassoConfig = {
  titleCase: boolean
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  picassoConfig?: PicassoConfig
): RenderResult => {
  const { rerender, ...rest } = render(
    <TestingPicasso titleCase={picassoConfig?.titleCase}>{ui}</TestingPicasso>,
    options
  )

  const customRerender: typeof rerender = newUi =>
    rerender(
      <TestingPicasso titleCase={picassoConfig?.titleCase}>
        {newUi}
      </TestingPicasso>
    )

  return { rerender: customRerender, ...rest }
}

export * from '@testing-library/react'
export { customRender as render, TestingPicasso }
export { getCheckpoints } from './get-checkpoints'
