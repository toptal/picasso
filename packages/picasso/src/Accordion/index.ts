import { OmitInternalProps } from '@toptal/picasso-shared'

import { Props } from './Accordion'

export { default } from './Accordion'
export type AccordionProps = OmitInternalProps<Props>

export { default as Details } from './Details'
export { default as Summary } from './Summary'
