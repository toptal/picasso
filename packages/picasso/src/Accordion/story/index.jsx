import { Accordion } from '../Accordion'
import { default as Details } from '../Details'
import { default as Summary } from '../Summary'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Accordion',
  `
    Accordions store information behind collapsible sections,
    allowing for more information to be stored in a limited amount of space.

    ${PicassoBook.createBaseDocsLink(
      'https://share.goabstract.com/636e68ee-2ef4-483d-bdaf-83aef2340477?collectionLayerId=aa49cc2c-6e4e-4d8d-aa3a-3def89e1837f&mode=design&present=true'
    )}
  `
)

page
  .createTabChapter('Props')
  .addComponentDocs({
    component: Accordion,
    name: 'Accordion',
  })
  .addComponentDocs({
    component: Details,
    name: 'Accordion.Details',
  })
  .addComponentDocs({
    component: Summary,
    name: 'Accordion.Summary',
  })

page
  .createChapter()
  .addExample('Accordion/story/Default.example.tsx', {
    title: 'Default',
    description:
      'Accordion is uncontrolled until the `expanded` prop is specified.',
  })
  .addExample('Accordion/story/Disabled.example.tsx', {
    title: 'Disabled',
    description:
      'Accordion ignores pointer events when the `disabled` prop is truthy.',
    takeScreenshot: false,
  })
  .addExample('Accordion/story/BorderedGroups.example.tsx', {
    title: 'Borders and Groups',
    description: 'Accordions have configurable borders and can be grouped',
  })
  .addExample('Accordion/story/Controlled.example.tsx', {
    title: 'Controlled',
    description: 'Accordion can be controlled via the `expanded` prop.',
    takeScreenshot: false,
  })
  .addExample('Accordion/story/CustomSummary.example.tsx', {
    title: 'Custom Summary',
    description:
      "Accordion's summary is customizable. It either can be passed as `children` or be an external custom component.",
    takeScreenshot: false,
  })
