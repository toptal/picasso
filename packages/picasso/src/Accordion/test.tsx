import React from 'react'
import { OmitInternalProps } from '@toptal/picasso-shared'
import { render, fireEvent, waitFor } from '@toptal/picasso/test-utils'

import Accordion, { Props } from './Accordion'

const DETAILS_TEXT =
  'A dog is a type of domesticated animal. Known for its loyalty and faithfulness, it can be found as a welcome guest in many households across the world.'
const SUMMARY_TEXT = 'What is a dog?'

const TestDetails = () => (
  <Accordion.Details data-testid='accordion-details'>
    {DETAILS_TEXT}
  </Accordion.Details>
)
const TestSummary = () => (
  <Accordion.Summary data-testid='accordion-summary'>
    {SUMMARY_TEXT}
  </Accordion.Summary>
)

const testIds = {
  emptyAccordionSummary: 'picasso-empty-accordion-summary',
}

const renderAccordion = (props?: Partial<OmitInternalProps<Props>>) =>
  render(
    <Accordion content={<TestDetails />} {...props}>
      <TestSummary />
    </Accordion>
  )

describe('Accordion', () => {
  it('renders collapsed by default', () => {
    const { container, queryByTestId, getByText, getByTestId } =
      renderAccordion()

    expect(queryByTestId('empty-accordion-summary')).toBeNull()

    expect(getByTestId('accordion-summary')).toBeVisible()
    expect(getByText(SUMMARY_TEXT)).toBeVisible()

    expect(getByTestId('accordion-details')).not.toBeVisible()
    expect(getByText(DETAILS_TEXT)).not.toBeVisible()

    expect(container).toMatchSnapshot()
  })

  it('renders empty summary when one is not provided', () => {
    const { getByTestId, queryByTestId } = render(
      <Accordion content={<TestDetails />} testIds={testIds} />
    )

    expect(getByTestId(testIds.emptyAccordionSummary)).toBeVisible()
    expect(queryByTestId('accordion-summary')).toBeNull()
    expect(getByTestId('accordion-details')).not.toBeVisible()
  })

  it('toggles', async () => {
    const handleChange = jest.fn()

    const { getByText, getByTestId } = renderAccordion({
      onChange: handleChange,
      expandIcon: <span data-testid='trigger' />,
    })

    fireEvent.click(getByTestId('accordion-summary'))
    await waitFor(() => expect(getByText(DETAILS_TEXT)).toBeVisible())

    fireEvent.click(getByTestId('trigger'))
    await waitFor(() => expect(getByText(DETAILS_TEXT)).not.toBeVisible())

    fireEvent.click(getByText(SUMMARY_TEXT))
    await waitFor(() => expect(getByText(DETAILS_TEXT)).toBeVisible())

    expect(handleChange).toHaveBeenCalledTimes(3)
  })

  it('renders disabled', async () => {
    const { container } = renderAccordion({ disabled: true })

    // MUI disabled state adds `pointer-events: none` style rule to the summary container.
    // It can't be tested programmatically `fireEvent` ignores this rule.
    expect(container).toMatchSnapshot()
  })

  it('renders expanded initially', async () => {
    const { getByText, getByTestId } = renderAccordion({
      defaultExpanded: true,
    })

    expect(getByTestId('accordion-details')).toBeVisible()
    expect(getByText(DETAILS_TEXT)).toBeVisible()

    fireEvent.click(getByTestId('accordion-summary'))

    await waitFor(() => {
      expect(getByTestId('accordion-details')).not.toBeVisible()
      expect(getByText(DETAILS_TEXT)).not.toBeVisible()
    })
  })

  it('renders custom icon when passed', () => {
    const { getByTestId, container } = renderAccordion({
      expandIcon: <span data-testid='custom-expand-icon' />,
    })

    expect(getByTestId('custom-expand-icon')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('passes styles correctly', () => {
    const { getByTestId } = renderAccordion({
      'data-testid': 'accordion',
      className: 'foobar',
      style: { display: 'table' },
    })

    const accordionContainer = getByTestId('accordion')

    expect(accordionContainer).toHaveStyle('display: table;')
    expect(accordionContainer.classList.contains('foobar')).toBeTruthy()
  })

  it('toggles when controlled', async () => {
    const { getByText, rerender } = render(
      <Accordion content={<TestDetails />} expanded={false}>
        <TestSummary />
      </Accordion>
    )

    expect(getByText(DETAILS_TEXT)).not.toBeVisible()

    rerender(
      <Accordion content={<TestDetails />} expanded>
        <TestSummary />
      </Accordion>
    )

    expect(getByText(DETAILS_TEXT)).toBeVisible()
  })
})
