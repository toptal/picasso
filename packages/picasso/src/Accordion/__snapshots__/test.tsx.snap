// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders collapsed by default 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiPaper-root MuiAccordion-root PicassoAccordion-root PicassoAccordion-bordersAll MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root PicassoAccordion-summary"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content PicassoAccordion-content"
        >
          <div
            class="PicassoAccordion-summaryWrapper"
            data-testid="accordion-summary"
          >
            What is a dog?
          </div>
          <div
            class="PicassoAccordion-expandIconAlignTop"
          >
            <button
              class="MuiButtonBase-root PicassoButton-small PicassoButtonAction-small PicassoButton-secondary PicassoButton-root PicassoButtonAction-root"
              data-component-type="button"
              tabindex="0"
              type="button"
            >
              <span
                class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content PicassoButtonAction-content"
              >
                <svg
                  class="PicassoSvgArrowDownMinor16-root PicassoButton-icon PicassoButtonAction-icon PicassoAccordion-expandIcon"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="m11.997 5.29.707.707-4 4-.707.707-.707-.707-4-4 .707-.707 4 4 4-4Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                class="MuiAccordionDetails-root PicassoAccordionDetails-root PicassoAccordion-details"
              >
                <div
                  class="PicassoAccordion-detailsWrapper"
                  data-testid="accordion-details"
                >
                  A dog is a type of domesticated animal. Known for its loyalty and faithfulness, it can be found as a welcome guest in many households across the world.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders custom icon when passed 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiPaper-root MuiAccordion-root PicassoAccordion-root PicassoAccordion-bordersAll MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root PicassoAccordion-summary"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content PicassoAccordion-content"
        >
          <div
            class="PicassoAccordion-summaryWrapper"
            data-testid="accordion-summary"
          >
            What is a dog?
          </div>
          <span
            class="PicassoAccordion-expandIcon"
            data-testid="custom-expand-icon"
          />
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                class="MuiAccordionDetails-root PicassoAccordionDetails-root PicassoAccordion-details"
              >
                <div
                  class="PicassoAccordion-detailsWrapper"
                  data-testid="accordion-details"
                >
                  A dog is a type of domesticated animal. Known for its loyalty and faithfulness, it can be found as a welcome guest in many households across the world.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders disabled 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiPaper-root MuiAccordion-root PicassoAccordion-root PicassoAccordion-bordersAll Mui-disabled MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        aria-disabled="true"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root PicassoAccordion-summary Mui-disabled Mui-disabled"
        role="button"
        tabindex="-1"
      >
        <div
          class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content PicassoAccordion-content"
        >
          <div
            class="PicassoAccordion-summaryWrapper"
            data-testid="accordion-summary"
          >
            What is a dog?
          </div>
          <div
            class="PicassoAccordion-expandIconAlignTop"
          >
            <button
              class="MuiButtonBase-root PicassoButton-small PicassoButtonAction-small PicassoButton-secondary PicassoButton-root PicassoButtonAction-root"
              data-component-type="button"
              tabindex="0"
              type="button"
            >
              <span
                class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content PicassoButtonAction-content"
              >
                <svg
                  class="PicassoSvgArrowDownMinor16-root PicassoButton-icon PicassoButtonAction-icon PicassoAccordion-expandIcon"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="m11.997 5.29.707.707-4 4-.707.707-.707-.707-4-4 .707-.707 4 4 4-4Z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              role="region"
            >
              <div
                class="MuiAccordionDetails-root PicassoAccordionDetails-root PicassoAccordion-details"
              >
                <div
                  class="PicassoAccordion-detailsWrapper"
                  data-testid="accordion-details"
                >
                  A dog is a type of domesticated animal. Known for its loyalty and faithfulness, it can be found as a welcome guest in many households across the world.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
